File: 백준/Bronze/10989. 수 정렬하기 3/수 정렬하기 3.java
import java.io.*;

public class Main {
    static int[] nums;

    private static void radixSort(int[] nums) {
        int[] result = new int[nums.length];
        int digit = 1;

        for (int i = 0; i < 5; i++) {
            int[] queue = new int[10];

            for (int num : nums) {
                queue[(num / digit) % 10]++;
            }

            for (int j = 1; j < 10; j++) {
                queue[j] += queue[j - 1];
            }

            for (int j = nums.length - 1; j >= 0; j--) {
                result[queue[nums[j] / digit % 10] - 1] = nums[j];
                queue[nums[j] / digit % 10]--;
            }

            for (int j = 0; j < nums.length; j++) {
                nums[j] = result[j];
            }

            digit *= 10;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        int n = Integer.parseInt(br.readLine());
        nums = new int[n];

        for (int i = 0; i < n; i++) {
            nums[i] = Integer.parseInt(br.readLine());
        }
        br.close();

        radixSort(nums);

        for (int i = 0; i < n; i++) {
            bw.write(nums[i] + "\n");
        }
        bw.flush();
        bw.close();
    }
}

File: 백준/Bronze/10989. 수 정렬하기 3/수 정렬하기 3.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

fun radixSort(nums: IntArray) {
    val result = IntArray(nums.size)
    var digit = 1

    repeat(5) {
        val queue = IntArray(10)

        for (num in nums) {
            queue[num / digit % 10]++
        }

        for (i in 1..9) {
            queue[i] += queue[i - 1]
        }

        for (i in nums.size - 1 downTo 0) {
            val digitValue = nums[i] / digit % 10
            result[queue[digitValue] - 1] = nums[i]
            queue[digitValue]--
        }

        for (i in nums.indices) {
            nums[i] = result[i]
        }

        digit *= 10
    }
}

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val bw = BufferedWriter(OutputStreamWriter(System.out))

    val n = br.readLine().toInt()
    val nums = IntArray(n) { br.readLine().toInt() }
    br.close()

    radixSort(nums)

    for (num in nums) {
        bw.write("$num\n")
    }
    bw.flush()
    bw.close()
}